service/tablet_allocator.cc

```diff
@@ -1021,21 +1000,60 @@ class load_balancer {
                 if (skip) {
                     lblogger.debug("Replication constraint check failed, unable to emit migration for replica ({}, {}) to co-habit the replica ({}, {})",
                         t2_id, src, t1_id, dst);
-                    return make_ready_future<>();
+                    return;
                 }
 
                 auto mig = create_migration_info(t2_id, src, dst);
-                auto mig_streaming_info = get_migration_streaming_info(_tm->get_topology(), *t2.info, mig);
+                auto mig_streaming_info = get_migration_streaming_info(_tm->get_topology(), *t2.desc.info, mig);
                 if (!can_accept_load(nodes, mig_streaming_info)) {
                     // FIXME: we can try another pair of non-colocated replicas of same sibling tablets.
                     lblogger.debug("Load limit reached, unable to emit migration for replica ({}, {}) to co-habit the replica ({}, {})",
                         t2_id, src, t1_id, dst);
-                    return make_ready_future<>();
+                    return;
                 }
                 apply_load(nodes, mig_streaming_info);
 
                 lblogger.info("Created migration for replica ({}, {}) to co-habit same shard as ({}, {})", t2_id, src, t1_id, dst);
                 plan.add(std::move(mig));
+            };
+
+            co_await tmap.for_each_sibling_tablets([&] (tablet_desc t1, std::optional<tablet_desc> t2_opt) -> future<> {
+                // Be optimistic about migrating tablets, as if they succeeded.
+                // Merge finalization will have to recheck that all sibling tablets are co-located.
+
+                if (!t2_opt) {
+                    on_internal_error(lblogger, format("Unable to find sibling tablet during co-location, with tablet count {}, for table {}",
+                                                       tmap.tablet_count(), table));
+                }
+                auto t2 = *t2_opt;
```

Comment:
nit: can be `auto&` to save a few cycles / memory accesses (not sure if it's optimized by the compiler or not).

Reply:
Also, maybe we could keep the data directly in `tablet_colocation_info` here and pass references to it to maybe_generate_colocation_plan, instead of the struct by value, like this:
```
            struct tablet_colocation_info {
                const tablet_desc& desc;
                tablet_replica_set sorted_set;
            };
...
    auto ti1 = tablet_colocation_info{ t1, get_replicas(t1) };
    auto ti2 = tablet_colocation_info{ *t2_opt, get_replicas(*t2_opt) };
```

Reply:
can be done in a follow up patch of course